### DISCUSSION TITLE
Failed to load model in virtual machine.

### DISCUSSION THREAD

**Discussion by fjea** [deepjavalibrary/djl#3872] [Q&A] [answered]
Hi guys:

I tried and failed to load a trained model using TensorFlow in VIRTUAL MACHINE with Linux OS  , but it works well in physical machine with the same Linux OS. 
Any help would be appreciated,thanks.

The specific details are as follows:

the dupmed err :
[hs_err_pid3013.log](https://github.com/user-attachments/files/31219290/hs_err_pid3013.log)

==================
jvm 2    | 2026-08-19 16:59:19.715 [WrapperSimpleAppMain] INFO  s.f.e.a.c.EqNoiseClassification.loadModel:61 - Building model loading standard...
jvm 2    | 2026-08-19 16:59:19.735 [WrapperSimpleAppMain] INFO  s.f.e.a.c.EqNoiseClassification.loadModel:73 - Model loading standard built successfully
jvm 2    | 2026-08-19 16:59:19.735 [WrapperSimpleAppMain] INFO  s.f.e.a.c.EqNoiseClassification.loadModel:75 - Model loading...
Loading:     100% |████████████████████████████████████████|
jvm 2    | 2026-08-19 16:59:23.794 [WrapperSimpleAppMain] INFO  a.d.u.Platform.detectPlatform:84 - Found matching platform from: jar:file:/home/ndws-eew-channel-1.0.2-SNAPSHOT/lib/tensorflow-native-cpu-2.16.1-linux-x86_64.jar!/native/lib/tensorflow.properties
**jvm 2    | #
jvm 2    | # A fatal error has been detected by the Java Runtime Environment:
jvm 2    | #**
jvm 2    | #  SIGILL (0x4) at pc=0x00007fbae62a154a, pid=3013, tid=3039
jvm 2    | #
jvm 2    | # JRE version: OpenJDK Runtime Environment BiSheng (17.0.19+13) (build 17.0.19+13)
jvm 2    | # Java VM: OpenJDK 64-Bit Server VM BiSheng (17.0.19+13, mixed mode, sharing, tiered, compressed class ptrs, g1 gc, linux-amd64)
jvm 2    | # Problematic frame:
jvm 2    | # C  [libtensorflow_framework.so.2+0x153454a]  google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)+0xba
jvm 2    | #
jvm 2    | # Core dump will be written. Default location: Core dumps may be processed with "/usr/lib/systemd/systemd-coredump %P %u %g %s %t %c %h" (or dumping to /home/ndws-eew-channel-1.0.2-SNAPSHOT/bin/core.3013)
jvm 2    | #
jvm 2    | # An error report file with more information is saved as:
jvm 2    | # /home/ndws-eew-channel-1.0.2-SNAPSHOT/bin/hs_err_pid3013.log
jvm 2    | #
jvm 2    | # If you would like to submit a bug report, please visit:
jvm 2    | #   https://gitcode.com/openeuler/bishengjdk-17/issues/
jvm 2    | # The crash happened outside the Java Virtual Machine in native code.
jvm 2    | # See problematic frame for where to report the bug.
jvm 2    | #

**Comment by 0xshubhs** [comment, marked as answer]
Not a DJL problem. Your VM's CPU is missing an instruction set TensorFlow needs.

```
#  SIGILL (0x4) at pc=0x00007fbae62a154a, pid=3013, tid=3039
#  C  [libtensorflow_framework.so.2+0x153454a]  google::protobuf::internal::OnShutdownRun(...)
```

SIGILL is illegal instruction, the CPU got an opcode it doesn't implement. TensorFlow's official x86_64 Linux builds have required AVX since 1.6 and `tensorflow-native-cpu-2.16.1-linux-x86_64` is one of them. Your physical box has AVX, the VM is presenting a generic CPU model that masks it.

`OnShutdownRun` is protobuf static init, so it dies during library load before your code runs. Matches your log, platform detected then gone.

Check on both machines:

```bash
grep -o -m1 'avx[0-9a-z_]*' /proc/cpuinfo | sort -u
```

Physical will show `avx`, `avx2`, `fma`. The VM won't.

Fix depends on hypervisor, all of them mean stop hiding host CPU features:

KVM/libvirt: `<cpu mode='host-passthrough'/>` or `host-model`
Proxmox: CPU type `host` instead of `kvm64`/`x86-64-v2`
VMware: check EVC isn't set below Sandy Bridge
Hyper-V: turn off processor compatibility mode

Needs a guest reboot, flags are fixed at boot.

If you can't change the CPU model there's no official non-AVX TF build. PyTorch and ONNX Runtime have lower baselines and would run.

**Reply by fjea** [reply]
Thank you so much, that really helps me. 
As you said, **"Your physical box has AVX, the VM is presenting a generic CPU model that masks it."**  my physical host  shows avx, avx2, fma and the VM won't. 
I changed the VM`S CPU work mode from “Compatibility Mode“ to “Host Matching Mode“， it shows AVXs , and then my trained model could be loaded and works well.
Thanks again！
